UserTools/VtxSeedFineGrid/VtxSeedFineGrid.cpp

```diff
@@ -0,0 +1,470 @@
+		for (int i = 0; i < 5; i++) {
+			for (int j = 0; j < 5; j++) {
+				for (int k = 0; k < 5; k++) {
+					Seed.SetZ(Center.at(l).Z() - 10 + 5 * i);
+					Seed.SetX(Center.at(l).X() - 10 + 5 * j);
+					Seed.SetY(Center.at(l).Y() - 10 + 5 * k);
+
+
+					//medianTime = this->GetMedianSeedTime(Seed);
+					thisFineSeed.SetVertex(Seed, medianTime);
+					thisFineSeed.SetDirection(SeedDir);
+					vSeedVtxList->push_back(thisFineSeed);
+
+				}
+			}
+		}
+	}
+}
+
+Direction VtxSeedFineGrid::findDirectionMRD() {
+	std::vector<BoostStore>* Tracks;
+	m_data->Stores["MRDTracks"]->Get("MRDTracks", Tracks);
+	m_data->Stores["MRDTracks"]->Get("NumMrdTracks", numtracksinev);
+
+	if (numtracksinev > 1) Log("Multiple tracks need work; just using first for now", v_debug, verbosity);
+	double gradx, grady, theta, phi;
+	Direction startVertex, endVertex, result;
+	BoostStore* thisTrack = &(Tracks->at(0));
+	
+	thisTrack->Get("VTrackGradient", gradx);
+	thisTrack->Get("HTrackGradient", grady);
+	theta = atan(grady / gradx);
+	phi = asin(pow((gradx*gradx + grady * grady), 0.5));
+	/*TRandom3 smear;
+	Direction vtxDir = fTrueVertex->GetDirection();
+	Direction result;
+	result.SetTheta(smear.Gaus(vtxDir.GetTheta(), 0.4));
+	result.SetPhi(smear.Gaus(vtxDir.GetPhi(), 0.4));*/
+	result.SetTheta(theta);
+	result.SetPhi(phi);
+
+	return result;
+}
+
+RecoVertex* VtxSeedFineGrid::FindSimpleDirection(RecoVertex* myVertex) {
+
+	/// get vertex position
+	double vtxX = myVertex->GetPosition().X();
+	double vtxY = myVertex->GetPosition().Y();
+	double vtxZ = myVertex->GetPosition().Z();
+	double vtxTime = myVertex->GetTime();
+
+	std::cout<<"Simple Direction Input Position: (" << vtxX << "," << vtxY << "," << vtxZ << ")\n";
+	// current status
+	// ==============
+	int status = myVertex->GetStatus();
+
+	/// loop over digits
+	/// ================
+	double Swx = 0.0;
+	double Swy = 0.0;
+	double Swz = 0.0;
+	double Sw = 0.0;
+	double digitq = 0.;
+	double dx, dy, dz, ds, px, py, pz, q;
+
+	RecoDigit digit;
+	for (int idigit = 0; idigit < fDigitList->size(); idigit++) {
+		digit = fDigitList->at(idigit);
+		if (digit.GetFilterStatus()) {
+			q = digit.GetCalCharge();
+			dx = digit.GetPosition().X() - vtxX;
+			dy = digit.GetPosition().Y() - vtxY;
+			dz = digit.GetPosition().Z() - vtxZ;
+			ds = sqrt(dx*dx + dy * dy + dz * dz);
+			px = dx / ds;
+			py = dx / ds;
+			pz = dz / ds;
+			Swx += q * px;
+			Swy += q * py;
+			Swz += q * pz;
+			Sw += q;
+		}
+	}
+
+	/// average direction
+	/// =================
+	double dirX = 0.0;
+	double dirY = 0.0;
+	double dirZ = 0.0;
+
+	int itr = 0;
+	bool pass = 0;
+	double fom = 0.0;
+
+	if (Sw > 0.0) {
+		double qx = Swx / Sw;
+		double qy = Swy / Sw;
+		double qz = Swz / Sw;
+		double qs = sqrt(qx*qx + qy * qy + qz * qz);
+
+		dirX = qx / qs;
+		dirY = qy / qs;   
+		pass = 1;
+	}
+
+	// set vertex and direction
+	// ========================
+	RecoVertex* newVertex = new RecoVertex(); // Note: pointer must be deleted by the invoker

```

Comment:
Question: why does this function return a pointer to an object on the heap? Can it not return an object? Can it not take an object reference and set that to the result? Can it wrap the returned pointer in a smart pointer to avoid the need for the caller to remember to delete it?